README
This software is for my research.  So, this may not be valuable to you guys, but disclose it as open source.


NEWS
* ribtools-0.1.0 (25 Aug 2010)
** initial version release


Local variables:
mode: outline
paragraph-separate: "[  ^L]*$"
coding: utf-8
End:


extract/mrt.h
/*_
 * Copyright 2010 Scyphus Solutions Co. Ltd.  All rights reserved.
 *
 * Authors:
 *      Hirochika Asai  <asai@scyphus.co.jp>
 */

/* $Id: mrt.h,v 0d6e5d383a92 2010/08/27 04:28:27 Hirochika $ */

#ifndef _MRT_H
#define _MRT_H

#include <stdint.h>
#include <stddef.h>

/*
 * http://tools.ietf.org/html/draft-ietf-grow-mrt-11
 */

#define MSG_BGP                 5       /* MRT only */
#define MSG_BGP4PLUS            9       /* MRT only */
#define MSG_TABLE_DUMP          12      /* dump bgp routes-mrt */
#define MSG_TABLE_DUMP_V2       13      /* RIPE RIS */
#define MSG_BGP4MP              16      /* dump bgp all */

#define BGP4MP_STATE_CHANGE     0
#define BGP4MP_MESSAGE          1
#define BGP4MP_ENTRY            2
#define BGP4MP_SNAPSHOT         4

/* Subtypes for TABLE_DUMP */
#define AFI_IP                  1
#define AFI_IP6                 2

/* Subtypes for TABLE_DUMP_V2 */
#define PEER_INDEX_TABLE        1
#define RIB_IPV4_UNICAST        2
#define RIB_IPV4_MULTICAST      3
#define RIB_IPV6_UNICAST        4
#define RIB_IPV6_MULTICAST      5
#define RIB_GENERIC             6

#define BGP_TYPE_OPEN           1
#define BGP_TYPE_UPDATE         2
#define BGP_TYPE_NOTIFICATION   3
#define BGP_TYPE_KEEPALIVE      4

#define BGP_ATTR_FLAG_EXTLEN    0x10

#define AS_SET                  1
#define AS_SEQUENCE             2
#define AS_CONFED_SEQUENCE      3
#define AS_CONFED_SET           4

#define BGP_ATTR_ORIGIN                 1
#define BGP_ATTR_AS_PATH                2
#define BGP_ATTR_NEXT_HOP               3
#define BGP_ATTR_MULTI_EXIT_DISC        4
#define BGP_ATTR_LOCAL_PREF             5
#define BGP_ATTR_ATOMIC_AGGREGATE       6
#define BGP_ATTR_AGGREGATOR             7
#define BGP_ATTR_COMMUNITIES            8
#define BGP_ATTR_ORIGINATOR_ID          9
#define BGP_ATTR_CLUSTER_LIST           10
#define BGP_ATTR_DPA                    11
#define BGP_ATTR_ADVERTISER             12
#define BGP_ATTR_RCID_PATH              13
#define BGP_ATTR_MP_REACH_NLRI          14
#define BGP_ATTR_MP_UNREACH_NLRI        15
#define BGP_ATTR_EXT_COMMUNITIES        16

/*
 * MRT header
 */
struct mrt_header {
    uint32_t time;
    uint16_t type;
    uint16_t subtype;
    uint32_t message_length;
};

/*
 * Table dump
 */
struct table_dump {
    uint16_t view_number;
    uint16_t sequence_number;
    union {
        uint32_t v4;
        uint32_t v6[4];
    } prefix;
    uint8_t prefix_length;
    uint8_t status;
    uint32_t originated_time;
    union {
        uint32_t v4;
        uint32_t v6[4];
    } peer_ip_addr;
    uint16_t peer_as;
    uint16_t attrlen;
};

/*
 * Peer index table
 */
struct peer_index_table_header {
    uint32_t collector_bgp_id;
    uint16_t view_name_length;
    unsigned char *view_name;   /* Pointer */
    uint16_t peer_count;
};
struct peer_index_table_record {
    char peer_type_ipv6;
    char peer_type_as4;
    uint32_t peer_bgp_id;
    union {
        uint32_t v4;
        uint32_t v6[4];
    } peer_ip_addr;
    union {
        uint16_t asn2;
        uint32_t asn4;
    } peer_as;
};
struct peer_index_table {
    struct peer_index_table_header header;
    struct peer_index_table_record *records;
};

/*
 * RIB unicast/multicast
 */
struct rib_header {
    uint32_t sequence_number;
    uint8_t prefix_length;
    uint8_t *prefix;
    uint16_t entry_count;
};
struct rib_entry {
    uint16_t peer_index;
    uint32_t originated_time;
    uint16_t attrlen;           /* Attribute length */
};
struct rib {
    struct rib_header header;
    struct rib_entry *entries;
};


#ifdef __cplusplus
extern "C" {
#endif

    struct mrt_header parse_mrt_header(unsigned char *);
    int parse_mrt_message(struct mrt_header *, unsigned char *, size_t);
    int parse_table_dump(struct mrt_header *, unsigned char *, size_t);
    int parse_table_dump_ip(struct mrt_header *, unsigned char *, size_t);
    int parse_table_dump_ip6(struct mrt_header *, unsigned char *, size_t);
    int parse_table_dump_v2(struct mrt_header *, unsigned char *, size_t);
    int
    parse_table_dump_v2_peer_index_table(
        struct mrt_header *, unsigned char *, size_t,
        struct peer_index_table **);
    int
    parse_table_dump_v2_rib(
        struct mrt_header *, unsigned char *, size_t,
        struct peer_index_table *);
    int parse_bgp_attribute(unsigned char *, size_t, int);
    int print_bgp_attribute(unsigned char *, size_t, int);

#ifdef __cplusplus
}
#endif

#endif /* _MRT_H */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */


extract/bsconv.c
/*_
 * Copyright 2010 Scyphus Solutions Co. Ltd.  All rights reserved.
 *
 * Authors:
 *      Hirochika Asai  <asai@scyphus.co.jp>
 */

/* $Id: bsconv.c,v 0d6e5d383a92 2010/08/27 04:28:27 Hirochika $ */

#include "bsconv.h"
#include <stdint.h>

/*
 * Convert byte stream to uint16_t
 */
uint16_t
bs2uint16(const unsigned char *bs, enum _endian endian)
{
    int i;
    uint16_t res;

    switch (endian) {
    case _ENDIAN_MACHINE:
        /* machine endian */
        res = *(uint16_t *)bs;
        break;
    case _ENDIAN_NETWORK:
    default:
        /* big endian */
        res = 0;
        for ( i = 0; i < 2; i++ ) {
            res <<= 8;
            res |= (uint16_t)bs[i];
        }
    }

    return res;
}

/*
 * Convert byte stream to uint32_t
 */
uint32_t
bs2uint32(const unsigned char *bs, enum _endian endian)
{
    int i;
    uint32_t res;

    switch (endian) {
    case _ENDIAN_MACHINE:
        /* machine endian */
        res = *(uint32_t *)bs;
        break;
    case _ENDIAN_NETWORK:
    default:
        /* big endian */
        res = 0;
        for ( i = 0; i < 4; i++ ) {
            res <<= 8;
            res |= (uint32_t)bs[i];
        }
    }

    return res;
}

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */


extract/extract-rib.c
/*_
 * Copyright 2009 WIDE Project. All rights reserved.
 *
 * Authors:
 *      Hirochika Asai  <panda@hongo.wide.ad.jp>
 */

/* $Id: extract-rib.c,v 0d6e5d383a92 2010/08/27 04:28:27 Hirochika $ */

#include "mrt.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

/*
 * Main
 */
int
main(int argc, const char *const argv[], const char *const envp[])
{
    unsigned char hbuf[12];
    int p;
    size_t nread;
    struct mrt_header header;
    unsigned char *body;

    /* 0-11:: header: time, type, subtype, packet length */
    /* 12-(packet length):: body */

    p = -1;
    for ( ;; ) {
        if ( feof(stdin) ) {
            break;
        }

        /* Read MRT header */
        nread = fread(hbuf, sizeof(char), sizeof(hbuf), stdin);
        /* Complete header? */
        if ( nread == 0 && feof(stdin) ) {
            /* End-of-file */
            break;
        } else if ( nread != 12 ) {
            /* Incomplete header */
            fputs("Incomplete header\n", stderr);
            exit(EXIT_FAILURE);
        }
        header = parse_mrt_header(hbuf);

        /* Read MRT message */
        body = malloc(header.message_length);
        if ( NULL == body ) {
            fputs("malloc(): Memory error\n.", stderr);
            exit(-1);
        }
        nread = fread(body, sizeof(char), header.message_length, stdin);
        if ( nread != header.message_length ) {
            /* Incomplete message */
            fputs("Incomplete message", stderr);
            exit(EXIT_FAILURE);
        }
        (void)parse_mrt_message(&header, body, nread);
        free(body);
    }

    return 0;
}


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */


extract/mrt.c
/*_
 * Copyright (c) 2010 Scyphus Solutions Co. Ltd.
 * Copyright (c) 2014 Hirochika Asai
 * All rights reserved.
 *
 * Authors:
 *      Hirochika Asai  <asai@jar.jp>
 */

#include "mrt.h"
#include "bsconv.h"
#include <stdio.h>
#include <stdlib.h>

/*
 * Parse MRT header
 */
struct mrt_header
parse_mrt_header(unsigned char *hbuf)
{
    struct mrt_header header;

    /* Parse header from the buffer */
    header.time = bs2uint32(hbuf, _ENDIAN_NETWORK);
    header.type = bs2uint16(hbuf+4, _ENDIAN_NETWORK);
    header.subtype = bs2uint16(hbuf+6, _ENDIAN_NETWORK);
    header.message_length = bs2uint32(hbuf+8, _ENDIAN_NETWORK);

    return header;
}

/*
 * Parse MRT message
 */
int
parse_mrt_message(struct mrt_header *header, unsigned char *body, size_t len)
{
    if ( MSG_TABLE_DUMP == header->type ) {
        parse_table_dump(header, body, len);
    } else if ( MSG_TABLE_DUMP_V2 == header->type ) {
        parse_table_dump_v2(header, body, len);
    } else {
        fputs("Unsupported type\n", stderr);
        return -1;
    }

    return 0;
}

/*
 * Parse TABLE_DUMP type
 */
int
parse_table_dump(struct mrt_header *header, unsigned char *body, size_t len)
{
    int ret;

    if ( AFI_IP == header->subtype ) {
        ret = parse_table_dump_ip(header, body, len);
    } else if ( AFI_IP6 == header->subtype ) {
        ret = parse_table_dump_ip6(header, body, len);
    } else {
        ret = -1;
    }

    return ret;
}

/*
 * Parse AFI_IPv4 of TABLE_DUMP type
 */
int
parse_table_dump_ip(struct mrt_header *header, unsigned char *body, size_t len)
{
    struct table_dump td;
    size_t rest;
    unsigned char *cptr;

    /* Check the length */
    if ( len < 22 ) {
        return -1;
    }

    cptr = body;
    rest = len;

    /* Get view # */
    td.view_number = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Get sequence number */
    td.sequence_number = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Get prefix */
    td.prefix.v4 = bs2uint32(cptr, _ENDIAN_NETWORK);
    cptr += 4;
    rest -= 4;

    /* Get prefix length */
    td.prefix_length = *cptr;
    cptr++;
    rest--;

    /* Get status */
    td.status = *cptr;
    cptr++;
    rest--;

    /* Get originated time */
    td.originated_time = bs2uint32(cptr, _ENDIAN_NETWORK);
    cptr += 4;
    rest -= 4;

    /* Get peer IP address */
    td.peer_ip_addr.v4 = bs2uint32(cptr, _ENDIAN_NETWORK);
    cptr += 4;
    rest -= 4;

    /* Get peer AS */
    td.peer_as = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Get attribute length */
    td.attrlen = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Check the length */
    if ( rest < td.attrlen ) {
        /* Length error */
        return -1;
    }

    /* IPv4 */
    printf("4 ");

    /* Prefix */
    printf("%d.%d.%d.%d/%d ",
           (td.prefix.v4 >> 24) & 0xff,
           (td.prefix.v4 >> 16) & 0xff,
           (td.prefix.v4 >> 8) & 0xff,
           td.prefix.v4 & 0xff,
           td.prefix_length);

    /* Print the peer IP address */
    printf("%d.%d.%d.%d ", (td.peer_ip_addr.v4 >> 24) & 0xff,
           (td.peer_ip_addr.v4 >> 16) & 0xff,
           (td.peer_ip_addr.v4 >> 8) & 0xff,
           td.peer_ip_addr.v4 & 0xff);

    /* Parse attribute */
    print_bgp_attribute(cptr, td.attrlen, 0);

    return 0;
}

/*
 * Parse AFI_IPv6 of TABLE_DUMP type
 */
int
parse_table_dump_ip6(struct mrt_header *header, unsigned char *body, size_t len)
{
    struct table_dump td;
    size_t rest;
    unsigned char *cptr;

    /* Check the length */
    if ( len < 46 ) {
        return -1;
    }

    cptr = body;
    rest = len;

    /* Get view # */
    td.view_number = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Get sequence number */
    td.sequence_number = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Get prefix */
    td.prefix.v6[0] = bs2uint32(cptr, _ENDIAN_NETWORK);
    td.prefix.v6[1] = bs2uint32(cptr+4, _ENDIAN_NETWORK);
    td.prefix.v6[2] = bs2uint32(cptr+8, _ENDIAN_NETWORK);
    td.prefix.v6[3] = bs2uint32(cptr+12, _ENDIAN_NETWORK);
    cptr += 16;
    rest -= 16;

    /* Get prefix length */
    td.prefix_length = *cptr;
    cptr++;
    rest--;

    /* Get status */
    td.status = *cptr;
    cptr++;
    rest--;

    /* Get originated time */
    td.originated_time = bs2uint32(cptr, _ENDIAN_NETWORK);
    cptr += 4;
    rest -= 4;

    /* Get peer IP address */
    td.peer_ip_addr.v6[0] = bs2uint32(cptr, _ENDIAN_NETWORK);
    td.peer_ip_addr.v6[1] = bs2uint32(cptr+4, _ENDIAN_NETWORK);
    td.peer_ip_addr.v6[2] = bs2uint32(cptr+8, _ENDIAN_NETWORK);
    td.peer_ip_addr.v6[3] = bs2uint32(cptr+12, _ENDIAN_NETWORK);
    cptr += 16;
    rest -= 16;

    /* Get peer AS */
    td.peer_as = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Get attribute length */
    td.attrlen = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    /* Check the length */
    if ( rest < td.attrlen ) {
        /* Length error */
        return -1;
    }

    /* IPv6 */
    printf("6 ");

    /* Prefix */
    printf("%.4x:%.4x:%.4x:%.4x:%.4x:%.4x:%.4x:%.4x/%d ",
           (td.prefix.v6[0] >> 16) & 0xffff,
           td.prefix.v6[0] & 0xffff,
           (td.prefix.v6[1] >> 16) & 0xffff,
           td.prefix.v6[1] & 0xffff,
           (td.prefix.v6[2] >> 16) & 0xffff,
           td.prefix.v6[2] & 0xffff,
           (td.prefix.v6[3] >> 16) & 0xffff,
           td.prefix.v6[3] & 0xffff,
           td.prefix_length);

    /* Print the peer IP address */
    printf("%.4x:%.4x:%.4x:%.4x:%.4x:%.4x:%.4x:%.4x ",
           (td.peer_ip_addr.v6[0] >> 16) & 0xffff,
           td.peer_ip_addr.v6[0] & 0xffff,
           (td.peer_ip_addr.v6[1] >> 16) & 0xffff,
           td.peer_ip_addr.v6[1] & 0xffff,
           (td.peer_ip_addr.v6[2] >> 16) & 0xffff,
           td.peer_ip_addr.v6[2] & 0xffff,
           (td.peer_ip_addr.v6[3] >> 16) & 0xffff,
           td.peer_ip_addr.v6[3] & 0xffff);

    /* Parse attribute */
    parse_bgp_attribute(cptr, td.attrlen, 0);

    return 0;
}

/*
 * Parse TABLE_DUMP_V2 type
 */
int
parse_table_dump_v2(struct mrt_header *header, unsigned char *body, size_t len)
{
    int ret;
    static struct peer_index_table *pit = NULL;

    if ( PEER_INDEX_TABLE == header->subtype ) {
        /* PEER_INDEX_TABLE */
        ret = parse_table_dump_v2_peer_index_table(header, body, len, &pit);
    } else if ( RIB_IPV4_UNICAST == header->subtype
                || RIB_IPV4_MULTICAST == header->subtype
                || RIB_IPV6_UNICAST == header->subtype
                || RIB_IPV6_MULTICAST == header->subtype ) {
        /* RIB */
        ret = parse_table_dump_v2_rib(header, body, len, pit);
    } else if ( RIB_GENERIC == header->subtype ) {
        fputs("Unsupported subtype\n", stderr);
        ret = -1;
    } else {
        ret = -1;
    }

    return ret;
}

/*
 * Parse TABLE_DUMP_V2's PEER_INDEX_TABLE
 */
int
parse_table_dump_v2_peer_index_table(
    struct mrt_header *header, unsigned char *body, size_t len,
    struct peer_index_table **pit)
{
    int i;
    size_t rest;
    unsigned char *cptr;
    struct peer_index_table_header pith;
    struct peer_index_table_record *rec;

    /* Check the length */
    if ( len < 8 ) {
        return -1;
    }

    /* Get collector BGP ID */
    pith.collector_bgp_id = bs2uint32(body, _ENDIAN_NETWORK);
    /* Get view name length */
    pith.view_name_length = bs2uint16(body+4, _ENDIAN_NETWORK);
    /* Check the header length */
    if ( len < 8 + pith.view_name_length ) {
        return -1;
    }
    /* Get view name */
    pith.view_name = body + 6;
    /* Get peer count */
    pith.peer_count = bs2uint16(body+6+pith.view_name_length, _ENDIAN_NETWORK);

    /* Get current pointer and rest */
    cptr = body + 8 + pith.view_name_length;
    rest = len - (8 + pith.view_name_length);

    /* Allocate for records */
    rec = malloc(sizeof(struct peer_index_table_record) * pith.peer_count);
    if ( NULL == rec ) {
        /* Memory error */
        return -1;
    }

    /* For all peers */
    for ( i = 0; i < pith.peer_count; i++ ) {
        /* Check the length */
        if ( rest < 5 ) {
            /* Length error */
            free(rec);
            return -1;
        }
        /* Get peer type */
        if ( *cptr & 1 ) {
            rec[i].peer_type_ipv6 = 1;
        } else {
            rec[i].peer_type_ipv6 = 0;
        }
        if ( *cptr & 2 ) {
            rec[i].peer_type_as4 = 1;
        } else {
            rec[i].peer_type_as4 = 0;
        }
        cptr++;
        rest--;
        /* Get peer BGP ID */
        rec[i].peer_bgp_id = bs2uint32(cptr, _ENDIAN_NETWORK);
        cptr += 4;
        rest -= 4;

        /*printf("%zu %zu %d\n", len, rest, rec[i].peer_type_ipv6);*/
        /* by type */
        if ( rec[i].peer_type_ipv6 ) {
            /* IPv6 */
            /* Check the length */
            if ( rest < 16 ) {
                /* Length error */
                free(rec);
                return -1;
            }
            /* Get IP address */
            rec[i].peer_ip_addr.v6[0] = bs2uint32(cptr, _ENDIAN_NETWORK);
            rec[i].peer_ip_addr.v6[1] = bs2uint32(cptr+4, _ENDIAN_NETWORK);
            rec[i].peer_ip_addr.v6[2] = bs2uint32(cptr+8, _ENDIAN_NETWORK);
            rec[i].peer_ip_addr.v6[3] = bs2uint32(cptr+12, _ENDIAN_NETWORK);
            cptr += 16;
            rest -= 16;
        } else {
            /* IPv4 */
            /* Check the length */
            if ( rest < 4 ) {
                /* Length error */
                free(rec);
                return -1;
            }
            /* Get IP address */
            rec[i].peer_ip_addr.v4 = bs2uint32(cptr, _ENDIAN_NETWORK);
            cptr += 4;
            rest -= 4;
        }
        /* by type */
        if ( rec[i].peer_type_as4 ) {
            /* 4-octet AS number */
            /* Check the length */
            if ( rest < 4 ) {
                /* Length error */
                free(rec);
                return -1;
            }
            /* Get AS number */
            rec[i].peer_as.asn4 = bs2uint32(cptr, _ENDIAN_NETWORK);
            cptr += 4;
            rest -= 4;
        } else {
            /* 2-octet AS number */
            /* Check the length */
            if ( rest < 2 ) {
                /* Length error */
                free(rec);
                return -1;
            }
            /* Get AS number */
            rec[i].peer_as.asn2 = bs2uint16(cptr, _ENDIAN_NETWORK);
            cptr += 2;
            rest -= 2;
        }
    }

    /* Set */
    *pit = malloc(sizeof(struct peer_index_table));
    if ( NULL == *pit ) {
        /* Memory error */
        free(rec);
        return -1;
    }
    (*pit)->header = pith;
    (*pit)->records = rec;

    return 0;
}

/*
 * Parse MRT TABLE_DUMP_V2's RIB
 */
int
parse_table_dump_v2_rib(
    struct mrt_header *header, unsigned char *body, size_t len,
    struct peer_index_table *pit)
{
    int i;
    unsigned char *cptr;
    size_t rest;
    int prefix_byte;
    struct rib rib;
    struct rib_header ribh;
    struct rib_entry *entries;
    int as4;

    if ( NULL == pit ) {
        /* Peer index table is not specified. */
        return -1;
    }

    rest = len;
    cptr = body;

    /* Check the length */
    if ( rest < 5 ) {
        return -1;
    }

    /* Get sequence number */
    ribh.sequence_number = bs2uint32(cptr, _ENDIAN_NETWORK);
    cptr += 4;
    rest -= 4;

    /* Get prefix length */
    ribh.prefix_length = *cptr;
    cptr++;
    rest--;

    if ( ribh.prefix_length > 0 ) {
        prefix_byte = (ribh.prefix_length-1)/8+1;
    } else {
        prefix_byte = 0;
    }

    /* Check length */
    if ( rest < prefix_byte ) {
        return -1;
    }

    /* Get prefix */
    ribh.prefix = cptr;
    cptr += prefix_byte;
    rest -= prefix_byte;

    /* Check length */
    if ( rest < 2 ) {
        return -1;
    }

    /* Get entry count */
    ribh.entry_count = bs2uint16(cptr, _ENDIAN_NETWORK);
    cptr += 2;
    rest -= 2;

    entries = malloc(sizeof(struct rib_entry) * ribh.entry_count);
    if ( NULL == entries ) {
        /* Memory error */
        return -1;
    }

    for ( i = 0; i < ribh.entry_count; i++ ) {
        /* Check the length */
        if ( rest < 8 ) {
            /* Length error */
            free(entries);
            return -1;
        }
        entries[i].peer_index = bs2uint16(cptr, _ENDIAN_NETWORK);
        cptr += 2;
        rest -= 2;
        entries[i].originated_time = bs2uint32(cptr, _ENDIAN_NETWORK);
        cptr += 4;
        rest -= 4;
        entries[i].attrlen = bs2uint16(cptr, _ENDIAN_NETWORK);
        cptr += 2;
        rest -= 2;

        /* Check the length */
        if ( rest < entries[i].attrlen ) {
            /* Length error */
            free(entries);
            return -1;
        }

        /* 4-octet? */
        if ( entries[i].peer_index >= pit->header.peer_count ) {
            fputs("Bad format\n", stderr);
            free(entries);
            return -1;
        }
        if ( pit->records[entries[i].peer_index].peer_type_as4 ) {
            as4 = 1;
        } else {
            as4 = 0;
        }

        if ( RIB_IPV4_UNICAST == header->subtype
             || RIB_IPV4_MULTICAST == header->subtype ) {
            uint8_t prefix[4];
            prefix[0] = ribh.prefix[0];
            prefix[1] = ribh.prefix[1];
            prefix[2] = ribh.prefix[2];
            prefix[3] = ribh.prefix[3];
            switch ( prefix_byte ) {
            case 0:
                prefix[0] = 0;
            case 1:
                prefix[1] = 0;
            case 2:
                prefix[2] = 0;
            case 3:
                prefix[3] = 0;
                break;
            }
            /* IPv4 */
            printf("4 ");

            /* Prefix */
            printf("%d.%d.%d.%d/%d ",
                   prefix[0], prefix[1], prefix[2], prefix[3],
                   ribh.prefix_length);

            /* Print the peer IP address */
            printf("%d.%d.%d.%d ",
                (pit->records[entries[i].peer_index].peer_ip_addr.v4 >> 24) & 0xff,
                (pit->records[entries[i].peer_index].peer_ip_addr.v4 >> 16) & 0xff,
                (pit->records[entries[i].peer_index].peer_ip_addr.v4 >> 8) & 0xff,
                pit->records[entries[i].peer_index].peer_ip_addr.v4 & 0xff);
        } else {
            uint8_t prefix[16];
            int j;
            for ( j = 0; j < 16; j++ ) {
                prefix[j] = ribh.prefix[j];
            }
            switch ( prefix_byte ) {
            case 0:
                prefix[0] = 0;
            case 1:
                prefix[1] = 0;
            case 2:
                prefix[2] = 0;
            case 3:
                prefix[3] = 0;
            case 4:
                prefix[4] = 0;
            case 5:
                prefix[4] = 0;
            case 6:
                prefix[4] = 0;
            case 7:
                prefix[4] = 0;
            case 8:
                prefix[4] = 0;
            case 9:
                prefix[4] = 0;
            case 10:
                prefix[4] = 0;
            case 11:
                prefix[4] = 0;
            case 12:
                prefix[4] = 0;
            case 13:
                prefix[4] = 0;
            case 14:
                prefix[4] = 0;
            case 15:
                prefix[4] = 0;
            case 16:
                prefix[4] = 0;
                break;
            }
            /* IPv6 */
            printf("6 ");

            /* Prefix */
            printf("%.2x%.2x:%.2x%.2x:%.2x%.2x:%.2x%.2x:%.2x%.2x:%.2x%.2x:%.2x%.2x:%.2x%.2x/%d ",
                   prefix[0], prefix[1], prefix[2], prefix[3],
                   prefix[4], prefix[5], prefix[6], prefix[7],
                   prefix[8], prefix[9], prefix[10], prefix[11],
                   prefix[12], prefix[13], prefix[14], prefix[15],
                   ribh.prefix_length);

            /* Print the peer IP address */
            printf("%.4x:%.4x:%.4x:%.4x:%.4x:%.4x:%.4x:%.4x ",
                   (pit->records[entries[i].peer_index].peer_ip_addr.v6[0] >> 16) & 0xffff,
                   pit->records[entries[i].peer_index].peer_ip_addr.v6[0] & 0xffff,
                   (pit->records[entries[i].peer_index].peer_ip_addr.v6[1] >> 16) & 0xffff,
                   pit->records[entries[i].peer_index].peer_ip_addr.v6[1] & 0xffff,
                   (pit->records[entries[i].peer_index].peer_ip_addr.v6[2] >> 16) & 0xffff,
                   pit->records[entries[i].peer_index].peer_ip_addr.v6[2] & 0xffff,
                   (pit->records[entries[i].peer_index].peer_ip_addr.v6[3] >> 16) & 0xffff,
                   pit->records[entries[i].peer_index].peer_ip_addr.v6[3] & 0xffff);
        }

        parse_bgp_attribute(cptr, entries[i].attrlen, as4);

        cptr += entries[i].attrlen;
        rest -= entries[i].attrlen;
    }

    rib.header = ribh;
    rib.entries = entries;
    free(entries);

    return 0;
}

/*
 * RFC4271  pp. 17
 */
int
parse_bgp_attribute(unsigned char *attr, size_t attrlen, int as4)
{
    size_t p;
    uint8_t flags;
    uint8_t type;
    uint16_t attriblen;
    unsigned char *attrib;
    int i;
    size_t x;
    int as_set;

    p = 0;
    while ( attrlen > p ) {
        flags = attr[p];
        type = attr[p+1];
        if ( flags & BGP_ATTR_FLAG_EXTLEN ) {
            /* Extended attribute length (16 bits) */
            attriblen = bs2uint16(attr+p+2, _ENDIAN_NETWORK);
            attrib = &attr[p+4];
            p = p + 4 + attriblen;
        } else {
            /* Normal attribute length (8 bits) */
            attriblen = attr[p+2];
            attrib = &attr[p+3];
            p = p + 3 + attriblen;
        }

        if ( BGP_ATTR_AS_PATH == type ) {
            /* seg_type = attrib[q]; */
            /* seg_length = attrib[q+1]; */
            /* seg_value = attrib[q+2:q+seg_length*as_size]; */
            x = 0;
            while ( x < attriblen ) {
                if ( 0 != x ) {
                    printf(" ");
                }
                if ( AS_SET == attrib[x] ) {
                    as_set = 1;
                } else {
                    as_set = 0;
                }
                if ( as_set ) {
                    printf("{");
                }
                for ( i = 0; i < attrib[x+1]; i++ ) {
                    if ( i != 0 ) {
                        if ( as_set ) {
                            printf(",");
                        } else {
                            printf(" ");
                        }
                    }
                    if ( as4 ) {
                        printf("%u", bs2uint32(&attrib[x+2+i*4], _ENDIAN_NETWORK));
                    } else {
                        printf("%u", bs2uint16(&attrib[x+2+i*2], _ENDIAN_NETWORK));
                    }
                }
                if ( as_set ) {
                    printf("}");
                }
                if ( as4 ) {
                    x += 2 + attrib[x+1] * 4;
                } else {
                    x += 2 + attrib[x+1] * 2;
                }
            }
            printf("\n");
        }
    }

    return 0;
}

/*
 * Temporal implementation
 * Excluding AS sets, and dividing by sequences
 */
int
print_bgp_attribute(unsigned char *attr, size_t attrlen, int as4)
{
    size_t p;
    uint8_t flags;
    uint8_t type;
    uint16_t attriblen;
    unsigned char *attrib;
    int i;
    size_t x;
    int f;
    int as_set;

    p = 0;
    while ( attrlen > p ) {
        flags = attr[p];
        type = attr[p+1];
        if ( flags & BGP_ATTR_FLAG_EXTLEN ) {
            /* Extended attribute length (16 bits) */
            attriblen = bs2uint16(attr+p+2, _ENDIAN_NETWORK);
            attrib = &attr[p+4];
            p = p + 4 + attriblen;
        } else {
            /* Normal attribute length (8 bits) */
            attriblen = attr[p+2];
            attrib = &attr[p+3];
            p = p + 3 + attriblen;
        }

        if ( BGP_ATTR_AS_PATH == type ) {
            /* seg_type = attrib[q]; */
            /* seg_length = attrib[q+1]; */
            /* seg_value = attrib[q+2:q+seg_length*as_size]; */
            x = 0;
            f = 0;
            while ( x < attriblen ) {
                if ( AS_SET == attrib[x] ) {
                    as_set = 1;
                } else {
                    as_set = 0;
                }
                if ( !as_set ) {
                    for ( i = 0; i < attrib[x+1]; i++ ) {
                        if ( f ) {
                            printf(" ");
                        }
                        if ( as4 ) {
                            printf("%u", bs2uint32(&attrib[x+2+i*4], _ENDIAN_NETWORK));
                        } else {
                            printf("%u", bs2uint16(&attrib[x+2+i*2], _ENDIAN_NETWORK));
                        }
                        f = 1;
                    }
                } else {
                    if ( f ) {
                        printf("\n");
                        f = 0;
                    }
                }
                if ( as4 ) {
                    x += 2 + attrib[x+1] * 4;
                } else {
                    x += 2 + attrib[x+1] * 2;
                }
            }
            if ( f ) {
                printf("\n");
                f = 0;
            }
        }
    }

    return 0;
}


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */
